src/software/thunderscope/common/proto_configuration_widget.py

```diff
@@ -58,9 +70,199 @@ def __init__(
+        """
+        load the protobuf from path_to_file to the variable self.proto_to_configure
+
+        :param path_to_file:
+        """
+        if not os.path.isfile(path_to_file):
+            self.proto_to_configure = ThunderbotsConfig()
+            self.proto_to_configure.sensor_fusion_config.friendly_color_yellow = (
+                self.is_yellow
+            )
+
+            self.build_proto(self.proto_to_configure)
+            self.save_current_config_to_file(self.path_to_file)
+            return
+
+        logging.info("Loading protobuf from file: {}".format(path_to_file))
+        with open(path_to_file, "rb") as f:
+            self.proto_to_configure = ThunderbotsConfig()
+            self.proto_to_configure.ParseFromString(f.read())
+
+            self.proto_to_configure.sensor_fusion_config.friendly_color_yellow = (
+                self.is_yellow
+            )
+
+    def send_proto_to_fullsystem(self):
+        """
+        Sending the default configuration protobufs to unix_full_system at startup.
+        Also updates the widget at the same time.
+        """
+        self.update_proto_from_file(self.path_to_file)
+        self.on_change_callback(None, None, self.proto_to_configure)
+        self.update_widget()
+
+    def save_proto_callback(self):
+        """
+        This is a callback for a button that save the current protobuf to disk!
+        """
+        try:
+            save_to_path, should_save = QFileDialog.getSaveFileName(
+                self,
+                "Select Protobufs",
+                self.get_default_savepath(),
+                options=QFileDialog.Option.DontUseNativeDialog,
+            )
+
+            if not should_save:
+                return
+
+
+            self.save_current_config_to_file(save_to_path)
+            self.update_widget()
+        except Exception:
+            logging.warning(
+                "cannot save configuration to {}".format(
+                    save_to_path
+                )
+            )
+
+    def save_current_config_to_file(self, path_to_file):
+        """
+        This save the self.proto_to_configure to path_to_file
+
+        :param path_to_file: the path to file that we are saving to.
+        """
+
+        logging.info("writing to file {}".format(path_to_file))
+        path_to_directory = os.path.dirname(path_to_file)
+        os.makedirs(path_to_directory, exist_ok=True)
+
+        with open(path_to_file, "wb") as f:
+            f.write(self.proto_to_configure.SerializeToString())
+
+    def load_proto_with_file_explorer(self):
+        """
+        loading the current protobuf through file explorer
+        """
+        try:
+            path_to_file, should_open = QFileDialog.getOpenFileName(
+                self,
+                "Select Protobufs",
+                self.get_default_savepath(),
+                options=QFileDialog.Option.DontUseNativeDialog,
+            )
+            if not should_open:
+                return
```

Comment:
What is the second return value of `getOpenFileName`? Couldn't find it in the docs: https://doc.qt.io/qtforpython-5/PySide2/QtWidgets/QFileDialog.html#PySide2.QtWidgets.PySide2.QtWidgets.QFileDialog.getOpenFileName

Reply:
We are not using PySlide but PyQt. That is the wrong documentation. 

This should parameter would return True if you have clicked the save button and False if you have clicked the cancel button. We want to not save when the user of clicked the cancel button, thus we early return in line 203.

Reply:
Oh okay that's interesting. I thought PySide and PyQt had 1:1 documentation with Qt, but I guess that's not the case for the entire API. 

I just tried it printing `should_open` and its and empty string if we cancel the dialogue, or its `should_open='All Files (*)'` if you do select a file. Either way the logic you have makes sense.